src/platforms/components/useSubmitEverything.tsx

```diff
@@ -0,0 +1,220 @@
+import useGuild from "components/[guild]/hooks/useGuild"
+import useCreateRole, {
+  RoleToCreate,
+} from "components/create-guild/hooks/useCreateRole"
+import useShowErrorToast from "hooks/useShowErrorToast"
+import useSubmit from "hooks/useSubmit"
+import useToast from "hooks/useToast"
+import { Requirement, RolePlatform } from "types"
+import { useFetcherWithSign } from "utils/fetcher"
+import preprocessRequirement from "utils/preprocessRequirement"
+
+type SubmitData =
+  /**
+   * If RoleToCreate is provided and `roleIds` is empty, a new role will be created
+   * with the associated rolePlatforms and requirements.
+   */
+  Partial<RoleToCreate> & {
+    /**
+     * Array of role IDs. If provided, rolePlatforms and requirements will be added to
+     * all specified roles. If not provided, RoleToCreate will be used to create a new
+     * role and associate them with it.
+     */
+    roleIds?: number[]
+
+    /**
+     * Array of requirements. If referenced in a rolePlatform, the requirement should
+     * have a temporal `id` generated with Date.now() A requirementIdMap will be
+     * created to match each temporalId-roleId pair with the actual backend-created ID
+     * after creation.
+     */
+    requirements: RoleToCreate["requirements"]
+
+    /**
+     * Array of rolePlatforms. These are created last, and their `roleId` fields will
+     * be filled based on the `roleIds` array or the new role ID if a new role was
+     * created.
+     */
+    rolePlatforms: RoleToCreate["rolePlatforms"]
+  }
+
+const useSubmitEverything = ({
+  methods,
+  onSuccess,
+}: {
+  methods: any
+  onSuccess: () => void
+}) => {
+  const toast = useToast()
+  const fetcherWithSign = useFetcherWithSign()
+  const { id: guildId } = useGuild()
+  const showErrorToast = useShowErrorToast()
+
+  const { onSubmit, isSigning, signLoadingText } = useCreateRole({
+    skipMutate: true,
+  })
+
+  const createRole = async (role: RoleToCreate) => {
+    try {
+      return await onSubmit(role)
+    } catch (error) {
+      showErrorToast("Failed to create role")
+      return null
+    }
+  }
+
+  const createRequirements = async (
+    requirements: Partial<Requirement>[],
+    roleIds: number[]
+  ) => {
+    const requirementIdMap = {}
```

Comment:
We should add a comment about the two dimensional object, because we can add the same requirement to different roles in the add reward flow, which means they will have different ids per role